### ISSUE TITLE
FR: Allow for seting `--reversed` globally in the config file

### ISSUE THREAD

**Issue by barakmich** [jj-vcs/jj#5203]
**Is your feature request related to a problem? Please describe.**

I set my `ui.default_command` to be `log --reversed` (even as [suggested in the documentation](https://jj-vcs.github.io/jj/latest/config/#default-command)) and I really like the reversed style. While there's also recent work related to reversing (eg, #5170), it would be nice to make reversing log order a global option, especially for when I'm running `jj log -r QUERY` and then forget to add `--reversed` every time.

**Describe the solution you'd like**

A configuration option for what seems to eventually be a family of `--reversed` flags, to invert their behavior. Some examples include:
```
ui.always_reversed = true
ui.top_to_bottom = true
```

Or if not globally, a consistent flag for each subcommand
```
log.default_reversed = true
```

(but it feels more `ui` to me :shrug:)

**Comment by PhilipMetzger** [comment]
I don't think this will be possible, see https://github.com/jj-vcs/jj/issues/1509. You can always use a custom log alias for it though.

**Comment by yuja** [comment]
And it's unlikely we'd want to see `op log` or `log -r 'all()'` in reverse order. I don't think `--reversed` by default would be worth the UI mess it would add.

**Comment by barakmich** [comment]
I do think it's useful in both those cases: `log -r 'all()'` is going to be a mess either way if it's really branchy or really long. Similar the `op log`

I think the underlying point to my request -- trying to take it from your perspective -- is that _if you're printing it straight to console_ having the most recent things, `op log` or `log`, is really handy to have the latest stuff the one or two lines above where you're typing. So I disagree that this is universally unlikely.

However, I can also see that, using the tool _in a script or pipeline to head/less or something_, it's easier/preferable to put the most recent at the top.

The semi-underlying issue being that, if you run `jj log -r 'all()'`, and it's too long, much like `git`, it will use a pager. Which is probably correct behavior, as well.

So we come back around to, maybe I'm just a minority, but, considering you advertise the option in the docs, it's really really helpful when taking a quick look around with `default-command` because it will print to console. So then it's confusing to folks getting used to the tool, and what one sees most frequently.

I'm closing the issue myself, but noting that while I think `jj` is great and going to do well, this felt really dismissive 😞

**Comment by ilyagr** [comment]
Just to expand on Philip's "alias" suggestion, that would be:

```toml
[aliases]
logr = ["log", "--reversed"]
evologr = ["evolog", "--reversed"]

[ui]
default-command = ["logr"]
```

Then, you can use `jj logr` instead of `jj log`. (I also think `jj op logr` will work because of how aliases are parsed, or it could have its own alias).

One upside of relying on aliases is that if you mention `jj log` to other people, it should work for them more or less the same as it works for you, and if you mention `jj logr` you will probably remember it's an alias the other person might not have. That might be the "UI mess" Yuya was referring to, I'm not sure.

**Comment by yuja** [comment]
> I do think it's useful in both those cases: `log -r 'all()'` is going to be a mess either way if it's really branchy or really long. Similar the `op log`

Just to be clear, I meant it would be unlikely because of performance reasons. `--reversed` graph can't be rendered without scanning all the graph nodes. It's only reasonably useful when the output is filtered by `-r` or the default revset.

And the UI cost we would have to pay is that we would need `--no-<flag>` for all boolean flags that can be flipped by configuration.

**Comment by arxanas** [comment]
Actually, I was going to change the default for git-branchless to have a `reversed` configuration option rather than a `--reversed` flag in https://github.com/arxanas/git-branchless/pull/1315#issuecomment-2132391412. (I should probably return to that PR... in fact, we could try to use a deprecation notice there to gain insight into workflows here, if we think it's valuable to do so.)

I do agree that reversing the output makes sense for a certain class of user's workflows (including me), although I'm sympathetic to the "UI cost" argument.

> I set my ui.default_command to be log --reversed (even as [suggested in the documentation](https://jj-vcs.github.io/jj/latest/config/#default-command))

[observation] I'm not sure if that suggestion was originally meant to advocate for reversing vs not reversing, vs just showing a helpful practical example of configuring your log output. I agree that the documentation sounds like it's recommending it as the default 👀.

> Just to be clear, I meant it would be unlikely because of performance reasons. --reversed graph can't be rendered without scanning all the graph nodes. It's only reasonably useful when the output is filtered by -r or the default revset.

One of the differences between git-branchless and jj that's relevant here is that git-branchless's log output by default tries to limit it to just "your" work, rather than all nodes, so it makes more sense to reverse the output via a configuration option.

---

> I'm closing the issue myself, but noting that while I think jj is great and going to do well, this felt really dismissive 😞

@barakmich Sorry to hear this. I don't think it was necessary to close the issue based on the discussion comments so far, which is perhaps a cultural difference for those inside and outside the project?

For this discussion, I'll reopen it for now to invite commentary from anyone else who may be interested in providing their feedback.
- If adopting the proposal in its current form, I think we would have to address the design concerns discussed above.
- Regardless, it would at least be good to hear if the `--reversed`-on-all-commands workflow is something that's currently unhandled for many users, even if we can't resolve the feature request in its current form.
- We can keep this open indefinitely if the fundamental workflow/UI issue remains unsolved, or migrate it to a Discussion or close it if we think we've achieved consensus.

I have found that many issues and proposals on fairly "basic" things get a lot of discussion about the topic itself, but also —

- relevant topics on different levels of abstractions
- reevaluations of whether the underlying version control concepts still apply to jj
- evaluations of feature and UI bloat, etc.
- discussion of whether the topic might be generalized into something in the distant future
- etc.

I think this is to be expected for a project at jj's stage, where not all of the design and UI principles have been fully established.
- Example: we will probably replace/supplant branches/bookmarks with some other design for managing related commits (such as ["topics"](https://github.com/jj-vcs/jj/issues/3402), and I myself have proposed a competing design). This means that a lot of simple requests for branches might devolve into discussions about whether we should implement a certain feature for branches at all, or other bikeshedding.

Was that in line with your expectations for the project? (You might also have just gotten unlucky with two immediate negative comments?) Do you think it's possible to convey those expectations to new users (perhaps via a comment in the feature request template)?

**Comment by yuja** [comment]
> One of the differences between git-branchless and jj that's relevant here is that git-branchless's log output by default tries to limit it to just "your" work, rather than all nodes,

That's also true for `jj`. The default `jj log` might be nice if rendered in reverse (depending on your preference.) The problem is that `jj log` can also be used as a command to query revisions of arbitrary length, and it's weird if `ui.reversed = true` were applied only to the default `jj log` (with no options.)

**Comment by Dianne087** [comment]
autotrace.set github because they dont appear in the plan. the dashes then. incicate

**Comment by Dianne087** [comment]
return.only me once githilub join specifies reverse.set.

**Comment by Dianne087** [comment]
%--.

**Comment by Dianne087** [comment]
Chapter ref --reverse

**Comment by Dianne087** [comment]
Abstract V protected on ()
My.Meth on

**Comment by Dianne087** [comment]
Abstract on
return github to solve this problem hidden my all tools use only his own strategy.
double( ), int( ) and long( )

**Comment by dotcarmen** [comment]
is there any sense in allowing revsets to be ordered? then a `reversed()` revset function could be defined

having ordered revsets would be presented similarly to when 2 branches with 2 independent histories are presented - eg:

```sh
$ jj log -r '(::kq | ::km) ~ root()' -T 'concat(change_id.shortest(), " is a commit")'
@  kq is a commit
○  m is a commit
│
~

○  km is a commit
│
~

```

with the proposed `reversed()` function, the output might look like:
```sh
$ jj log -r '(reversed(::kq) | ::km) ~ root()' -T 'concat(change_id.shortest(), " is a commit")'
~
│
○  m is a commit
@  kq is a commit

○  km is a commit
│
~
```

when there's a conflict in ordering, `jj log` should render separate graphs as above (note the absent `~ root()` below):
```sh
$ jj log -r 'reversed(::kq) | ::km' -T 'concat(change_id.shortest(), " is a commit")'
◆  z is a commit
○  m is a commit
@  kq is a commit

○  km is a commit
◆  z is a commit
```

**Comment by yuja** [comment]
Mercurial supports sorting in revsets, but it was source of random bugs. We have to track which side of `x&y` dominates the order, whether the sub expression can be computed lazily, etc. 

We could add an extended expression (similar to #5307) that can only be used as the outermost expression, though.
